cpp/src/arrow/types/primitive.h

```diff
@@ -261,10 +264,10 @@ typedef NumericBuilder<DoubleType> DoubleBuilder;
 
 class BooleanArray : public PrimitiveArray {
  public:
-  using PrimitiveArray::PrimitiveArray;
-
   BooleanArray(int32_t length, const std::shared_ptr<Buffer>& data,
       int32_t null_count = 0, const std::shared_ptr<Buffer>& null_bitmap = nullptr);
+  BooleanArray(const TypePtr& type, int32_t length, const std::shared_ptr<Buffer>& data,
```

Comment:
What is the benefit of passing a TypePtr? This array should always have the same type?